pylint/lint/pylinter.py

```diff
@@ -1019,6 +1053,8 @@ def check(self, files_or_modules: Union[Sequence[str], str]) -> None:
                 DeprecationWarning,
             )
             files_or_modules = (files_or_modules,)  # type: ignore[assignment]
+        if self.config.recursive:
```

Comment:
It seems to happen before in the initialization but should we make recursive true, if pylint is called without directory (``pylint --rc-file=...``), or with ``.`` (``pylint . --rc-file=...``) and if ``__init__.py``does not exists in the current directory ? I think the intent to use recursive is clear for those two as it would only work if there was an ``__init__.py`` present. So we could fix it right now (before releasing pylint 3.0). On the other hand modifying behavior based on other argument than the one controlling the behavior feels wrong. @DanielNoord, thought ?

Reply:
> It seems to happen before in the initialization but should we make recursive true, if pylint is called without directory (`pylint --rc-file=...`), or with `.` (`pylint . --rc-file=...`) and if `__init__.py`does not exists in the current directory ?

I don't think `pylint --rc-file=...` should be allowed. It should either be `pylint . --rc-file=...` or with another directory.
We could discuss allowing `pylint --rc-file=...` if we add support for a `files` option in `rc-file`, but running `pylint` without any clear intent on what directory to lint wouldn't be good imo. How would we display the help message then?

> I think the intent to use recursive is clear for those two as it would only work if there was an `__init__.py` present. So we could fix it right now (before releasing pylint 3.0). On the other hand modifying behavior based on other argument than the one controlling the behavior feels wrong. @DanielNoord, thought ?

I agree that `recursive=y` should be set when called with `.` as directory. The same probably goes for missing `__init__.py`. However, those would be breaking changes and I wonder how much depends on this currently not being supported. I'd be hesitant to change this in a minor version.

Reply:
> I don't think pylint --rc-file=... should be allowed. 


pytest does this I think it's a nice default value. It was asked [here and has currently 12 upvotes](https://github.com/PyCQA/pylint/issues/352#issuecomment-381300520)


> How would we display the help message then?


`pylint --help`, or `pylint -h`,  which I think is the expectation for CLI tool. (also ``pylint --long-help`` because we're special :smile:)


> I agree that recursive=y should be set when called with . as directory. The same probably goes for missing __init__.py. 


I meant both `.` and no `__init__.py` in the current directory at the same time. In this configuration the current result would be a fatal and nothing working so I guess it would not be a real breaking change (although there's probably someone somewhere who check for pylint exiting with a fatal instead of checking for an `__init__.py` :smile: ).

You made me realize that in fact we can make recursive true when there is no `__init__.py` in any directory we lint and it does not matter if it's `.` the argument that it was not working at all before is still true. But changing the recusrive option for `.` would be a breaking change as pylint could be working before if there was an `__init__.py` in the top level directory, but would start to lint new files unexpectedly.

Reply:
> pytest does this I think it's a nice default value. It was asked [here and has currently 12 upvotes](https://github.com/PyCQA/pylint/issues/352#issuecomment-381300520)

`Mypy` also allows this but only when you set `files` in your config. I really think we shouldn't allow this without setting `files` first. 

> `pylint --help`, or `pylint -h`, which I think is the expectation for CLI tool. (also `pylint --long-help` because we're special 😄)

My expectation is that just running the command also gives a short help message. I know so many CLI tool that do this.

Reply:
> My expectation is that just running the command also gives a short help message. I know so many CLI tool that do this.

Yes, this is what happen when the tool can't work without providing arguments. For example `ls` or `pwd`, `pytest` or `flake8` can run without args so they do. `mv` or `scp` can't, so they display the help (or something telling to how to display the help in the case of mv):
```
$scp
usage: scp [-346BCpqrTv] [-c cipher] [-F ssh_config] [-i identity_file]
            [-J destination] [-l limit] [-o ssh_option] [-P port]
            [-S program] source ... target
```

Imo, pylint should work without any argument (like flake8). But I wonder why mypy do not though. Do you know the rational behind it ?

Reply:
I looked in their repository for issues with "run", "parameter", "arguments" etc. but couldn't find anything.

I mean this boils down to one issue: can `pylint` run without arguments? I'd say no, as we need to know which directory to lint. `mypy` has answered that question similarly and says "as long as you don't provide an `files` argument I can't run". `flake8` has responded by assuming that if there is no `files` argument the user probably wants to run on `curdir`. Both are valid, but I like `mypy`'s response more as it makes no assumptions about what the user intends while still offering the desired behaviour (you just need to set `files` once in your `pyproject.toml`).

Besides this discussion I would vote against this as it breaks a core behaviour of `pylint`, namely how it is invoked. I don't think we should do such changes if there are no large benefits. In this case the benefit would be allowing to write `pylint` instead of `pylint .`. You know how much I like breaking changes and `DeprecationWarnings` (😄) but for me this is not enough to do this.
_Setting `--recursive=y` as the default is different to me as it is so much longer compared to writing ` .`._

Reply:
> I looked in their repository for issues with "run", "parameter", "arguments" etc. but couldn't find anything.

Yeah neither do I.

> I like mypy's response more as it makes no assumptions about what the user intends while still offering the desired behavior

I think to default to current directory with recursive search is rather intuitive (and popular as it gathered more than 10 up votes in the original issue). But let's defer it to pylint 3.0 as it's only slightly related to the original problem and is basically adding a default value which we can do it easily when a consensus is reached. I'm going to create another issue and we'll be able to direct the original issue watchers to it and get their opinion.

Are we agreeing on recursive being activated automatically if there is no `__init__.py` in the given directory ?

Reply:
Yes I think so!

Reply:
> > I looked in their repository for issues with "run", "parameter", "arguments" etc. but couldn't find anything.
> 
> Yeah neither do I.
> 
> > I like mypy's response more as it makes no assumptions about what the user intends while still offering the desired behavior
> 
> I think to default to current directory with recursive search is rather intuitive (and popular as it gathered more than 10 up votes in the original issue). But let's defer it to pylint 3.0 as it's only slightly related to the original problem and is basically adding a default value which we can do easily when a consensus is reached. I'm going to create another issue and we'll be able to direct the original issue watchers to it and get their opinion.
> 
> Are we agreeing on recursive being activated automatically if there is no `__init__.py` in the given directory ?

If I read it correctly, we are not going to change default behaviour of pylint? Is there anything pending it this PR?

Reply:
I would say we change the behavior a little: when the user want to lint a directory without an ``__init__.py`` inside it, we make ``recursive`` to be true. I think the tests needs to be adapted for a directory without init file and one test for a file that does not exists could be added.

Reply:
It means 
1. executing command `pylint` inside a directory `spam` without `__init__.py` file will trigger also recursive behaviour
2. executing command `pylint .` inside a directory `spam` without `__init__.py` file will trigger also recursive behaviour 

Is that correct?

Reply:
For 1, no, it should show the help like before, we did not reach a consensus on this with @DanielNoord  and I'm going to create an issue to settle it by popular vote, once we merge this one. (By the way, what do you think of defaulting to "." ? :) )

For 2. yes exactly.

Reply:
Just for myself but perhaps helpful for others. I think for now this is the summary:
For `2.x`:
When in `spam` without `__init__.py`:
`pylint .` > Recursive
`pylint` > Exit without doing anything.
`pylint ../spam` > ??? (Recursive I guess)?

When in `spam` with `__init__.py`:
`pylint .` > not recursive
`pylint` >  Exit without doing anything.
`pylint ../spam` > ??? (Not recursive I guess)?

When in parent without `__init__.py` of `spam` without `__init__.py`:
`pylint .` > Recursive, finds `spam`
`pylint spam` > Recursive
`pylint` > Exit without doing anything.
`pylint ../.` > ??? (Recursive I guess)?

When in parent without `__init__.py` of `spam` with `__init__.py`:
`pylint .` > Recursive, finds `spam`
`pylint spam` > Not recursive
`pylint` > Exit without doing anything.
`pylint ../.` > ??? (Recursive I guess)?

-----------------
For `3.0`. In bold is those that changed compared to `2.x`:
When in `spam` without `__init__.py`:
`pylint .` > Recursive
`pylint` > **TBD.**
`pylint ../spam` > **Recursive**

When in `spam` with `__init__.py`:
`pylint .` > **Recursive**
`pylint` > **TBD.**
`pylint ../spam` > **Recursive**

When in parent without `__init__.py` of `spam` without `__init__.py`:
`pylint .` > Recursive, finds `spam`
`pylint spam` > Recursive
`pylint` > **TBD.**
`pylint ../.` > **Recursive**

When in parent without `__init__.py` of `spam` with `__init__.py`:
`pylint .` > Recursive, finds `spam`
`pylint spam` > **Recursive**
`pylint` > **TBD**
`pylint ../.` > **Recursive**

Reply:
> ??? (Recursive I guess)?

I would say those should indeed be recursive.

Reply:
Last commit implements default recursive mode when `.` directory is passed to pylint.

Reply:
Let's leave this comment as unresolved as it is quite important for future people looking back at this PR and it might get easily lost.